src/OpenTracing/SpanReference.php

```diff
@@ -2,20 +2,93 @@
 
 namespace OpenTracing;
 
-interface SpanReference
+use OpenTracing\Exceptions\InvalidReferenceArgument;
+
+final class SpanReference
 {
+    const CHILD_OF = 'child_of';
+    const FOLLOWS_FROM = 'follows_from';
+
     /**
-     * @return bool
+     * @var SpanContext
      */
-    public function isTypeChildOf();
+    private $context;
 
     /**
-     * @return bool
+     * @var string
+     */
+    private $type;
+
+    private function __construct(SpanContext $context, $type)
+    {
+        $this->context = $context;
+        $this->type = $type;
+    }
+
+    /**
+     * @param SpanContext|Span $context
+     * @param string $type
+     * @throws InvalidReferenceArgument on empty type
+     * @return SpanReference when context is invalid
+     */
+    public static function create($context, $type)
+    {
+        if (empty($type)) {
+            throw InvalidReferenceArgument::emptyType();
+        }
+
+        return new self(self::extractContext($context), $type);
+    }
+
+    /**
+     * @param SpanContext|Span $context
+     * @throws InvalidReferenceArgument when context is invalid
+     * @return SpanReference
+     */
+    public static function createAsChildOf($context)
+    {
+        return new self(self::extractContext($context), self::CHILD_OF);
```

Comment:
I don't see the point of these factory functions, `new SpanReference($context, SpanReference::CHILD_OF)` is not longer and easier to reason about than `SpanReference::createAsChildOf()`. The latter suggests there is some heavy logic involved (otherwise it wouldn't need a factory function), but in reality it just sets a property.